Bug 1930985 - Implement search engine order handling on the Rust base…

mandysGit opened the pull request:
…d SearchEngineSelector.

### Pull Request checklist ###
<!-- Before submitting the PR, please address each item -->
- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [ ] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due diligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

mandysGit commented:
This PR is still work in progress and there will be more changes soon. 
I need a first pass review, please. 

I've added the implemention for `orderHint` property. I included setting the sort priority and sorting the engines orders. 

The part I need some feedback on right now is the first test I wrote, and of course everything else if there's any suggestions. 

I converted this test into Rust in the `selector.rs`: 

https://searchfox.org/mozilla-central/source/toolkit/components/search/tests/xpcshell/test_engine_selector_engine_orders.js#163-173

And I'm copying over this engine config from the top of the file:

https://searchfox.org/mozilla-central/source/toolkit/components/search/tests/xpcshell/test_engine_selector_engine_orders.js#15-43

As I was following the test structure in `selector.rs` I noticed the config and expected_engines parts are duplicated and very long. The way we've done it in Desktop is that we hadn't need to write all of the properties in. I wanted some feedback if I should continue the tests in this format even though they're quite long and has lots of duplication? 

My aim is to continue writing [tests for starts_with and defaults engines](https://searchfox.org/mozilla-central/source/toolkit/components/search/tests/xpcshell/test_engine_selector_engine_orders.js#208-240). But that would mean I'll need to create another json config and another expected_engines list -- which I think is fine to do but there's a lot of duplication. 



Curious to know what your thoughts on this are?

Standard8 commented on components/search/src/configuration_types.rs:
```diff
@@ -240,11 +240,28 @@ pub(crate) struct JSONDefaultEnginesRecord {
+    /// The order of the engines for the associated environment. If engines are
+    /// present for the user but not included in this list, they will follow
+    /// after the ones in this list in alphabetical order. If an individual entry
+    /// is suffixed with a star, matching is applied on a \"starts with\" basis.
```
nit: Shouldn't need the backslashes here.

Standard8 commented on components/search/src/filter.rs:
```diff
@@ -241,6 +266,49 @@ fn find_engine_with_match(
+) -> std::cmp::Ordering {
+    let b_index = get_priority(b, default_engine_id, default_private_engine_id);
+    let a_index = get_priority(a, default_engine_id, default_private_engine_id);
+    b_index.cmp(&a_index)
```
There's a piece missing here, which the engine selector on desktop doesn't do - the items which aren't default/have an order hint set, need to be sorted alphabetically by their display name.

This is the full function we use: https://searchfox.org/mozilla-central/rev/5b061cdc4d40d44988dc61aa941cfbd98e31791f/toolkit/components/search/SearchUtils.sys.mjs#362-440

One thing I'm not sure about is how we'll handle the locale-based sorting like we do on desktop. I haven't looked into the options for sort functions on Rust.

Standard8 commented on components/search/src/filter.rs:
```diff
@@ -221,7 +235,18 @@ fn find_engine_id(engines: &[SearchEngineDefinition], engine_id: String) -> Opti
 }
 
-fn find_engine_with_match(
+fn set_engine_order(engines: &mut [SearchEngineDefinition], ordered_engines: &[String]) {
```
I'm tempted to say we could put the sort related functions into a separate module, which might make it easier to write unit tests for them (and smaller module sizes), what do you think?

mandysGit commented:
I addressed the review comments above. Please resolve them if you think they're done. 

I added unit tests for sort function and I kept the larger tests in `selector.rs`. The tests in `selector.rs` have been refactored so it's a bit easier to read because I extract out only the relevant pieces which are the engine ids for comparing. 

I'm thinking maybe we can refactor the rest of the tests with a similar helper function and in combination with the builder pattern you've been working on. 

Lastly, I'm a bit worried the git checks won't pass since it didn't pass last time. I'm not too sure how to run the lint locally or fix them.

Standard8 commented on components/search/src/sort_helpers.rs:
```diff
@@ -0,0 +1,249 @@
+        }
+    }
+
+    #[test]
```
We should add a unit test here for the "starts with" match. We then wouldn't necessarily need the test in selector.rs, since we know that part is being tested here.

mandysGit replied:
I've added a unit test in `sort_helpers.rs` to test for the "starts with" match. 

I'm hesitant to remove the test in selectors.rs because it does more than just verify the "starts with" match. It also ensures the correct filtering based on environment, verifying that it finds the correct `wiki-uk` or `wiki-ca` engines based on locale and region when given `wiki*` as the engine_id to match. 

Whereas the "starts with" match in sort_helpers, I can only test it finds the first engine that starts with `wiki` so it will always match `wiki-ca` and and ignore `wiki-uk`.

Standard8 reviewed (changes requested):
Thank you for the updates, this is looking a lot nicer. I've just left a few comments in-line to consider.

mandysGit commented on components/search/src/sort_helpers.rs:
```diff
@@ -0,0 +1,249 @@
+                "a-engine".to_string(),
+            ],
+        );
+        let expected_order = vec![
```
I agree. This was confusing. 
I made some changes to both the tests for set engine orders: 

- I used `expected_order_hints` and `actual_order_hints`
- changed the engine ids to not use `a`, `b,`, `c`. Instead I used: `last-engine`, `secondary-engine`, `primary-engine` and `no-order-hint-engine` to make it more clear we're not expecting alphabetical sorting here
- Changed the test descriptions 
- extracted out the ordered_engines_lists as variable to pass into `set_engine_order`. I felt like, when it was passed in as a vector literal, it wasn't clear that it's the ordered list.

Standard8 reviewed (approved):
Thank you, this looks good. Just the headers in the new file to sort out.
